### PR TITLE
PHPLIB-493: Unified test runner POC

### PR DESCRIPTION
**Pull request by jmikola**
https://jira.mongodb.org/browse/PHPLIB-493

### CODE REVIEW

File: tests/UnifiedSpecTests/Constraint/IsBsonType.php [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+
+    private function doMatches($other) : bool
+    {
+        switch ($this->type) {
+            case 'double':
+                return is_float($other);
+            case 'string':
+                return is_string($other);
+            case 'object':
+                return is_object($other) && (! $other instanceof BSONArray);
```
Comment (alcaeus) [inline tests/UnifiedSpecTests/Constraint/IsBsonType.php:82]: Any BSON type that is represented by an object, including a 64-bit integer on a 32-bit platform will return true for this check. Is that intended? I'm wondering if this check should not only exclude `BSONArray`, but also exclude any `MongoDB\BSON\Type` instance to avoid false positives?

Edit: noticed that I built the original code in `DocumentMatchConstraint`, so I'll happily take the blame for missing any type overlaps while writing this.

File: tests/UnifiedSpecTests/Constraint/Match.php [hunk 2]
```diff
@@ -0,0 +1,470 @@
...
+        }
+
+        if (! $success) {
+            $this->fail($other, $description, $this->lastFailure);
+        }
+    }
+
+    private function assertEquals($expected, $actual, string $keyPath)
+    {
+        $expectedType = is_object($expected) ? get_class($expected) : gettype($expected);
```
Comment (alcaeus) [inline tests/UnifiedSpecTests/Constraint/Match.php:97]: FYI, PHP 8 introduces `get_debug_type` to avoid these checks. If you want, we can introduce the [symfony/polyfill](https://github.com/symfony/polyfill) package to provide this for us.

File: tests/UnifiedSpecTests/Constraint/Match.php [hunk 3]
```diff
@@ -0,0 +1,470 @@
...
+     */
+    private function assertMatches($expected, $actual, bool $ignoreExtraKeys, $keyPath = '')
+    {
+        if (! $expected instanceof BSONDocument && ! $expected instanceof BSONArray) {
+            $this->assertEquals($expected, $actual, $keyPath);
+
+            return;
+        }
+
+        if ($expected instanceof BSONArray) {
```
Comment (alcaeus) [inline tests/UnifiedSpecTests/Constraint/Match.php:148]: Maybe we can extract some of this logic to their own methods or even own constraint classes (if it wouldn't lead to too much code duplication)? As is, the method is a little hard to follow since it covers so many different use-cases.

File: tests/UnifiedSpecTests/Constraint/IsBsonType.php [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+use function range;
+use function sprintf;
+use const PHP_INT_SIZE;
+
+final class IsBsonType extends Constraint
+{
+    use ConstraintTrait;
+
+    private static $knownTypes = [
+        'double' => 1,
```
[the remark this answers is not in the document]
Comment (jmikola) [reply tests/UnifiedSpecTests/Constraint/IsBsonType.php:40]: This is just for `isset`. I figured it was cheaper than `in_array()` or `array_fill_keys()` with `isset()`.

File: tests/UnifiedSpecTests/Constraint/IsBsonType.php (hunk 1 above)
Comment (jmikola) [reply tests/UnifiedSpecTests/Constraint/IsBsonType.php:82]: Ah, good catch. I did indeed copy this from DocumentMatchConstraint. It was just the array detection that I rewrote.

I'll break this up into a separate `isObject()` method to cover the cases.

File: tests/UnifiedSpecTests/Constraint/Match.php (hunk 2 above)
Comment (jmikola) [reply tests/UnifiedSpecTests/Constraint/Match.php:97]: Great idea. Opened [PHPLIB-588](https://jira.mongodb.org/browse/PHPLIB-588).

File: tests/UnifiedSpecTests/Constraint/Match.php (hunk 3 above)
Comment (jmikola) [reply tests/UnifiedSpecTests/Constraint/Match.php:148]: Aye, I did intend on breaking this up into separate methods for the array/object cases.

I think there may be diminishing returns on turning everything into a constraint, since our only use case is calling `matches()` to get a boolean result. Ideally, we could rely on the constraint's own semantic error and concatenate our field path info from MatchConstraint, but I didn't want to waste more time trying to figure out how PHPUnit's Constraint API works (and perhaps this is made more difficult due to the BC brdige we're using).

File: tests/UnifiedSpecTests/Constraint/IsBsonType.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (jmikola) [reply tests/UnifiedSpecTests/Constraint/IsBsonType.php:40]: Let me know if you'd prefer I change this. Otherwise I'll defer to you to resolve.

File: tests/UnifiedSpecTests/Constraint/Match.php (hunk 3 above)
Comment (jmikola) [reply tests/UnifiedSpecTests/Constraint/Match.php]: Just pushed fixes and tests for both constraint classes, so PTAL.

File: tests/UnifiedSpecTests/Constraint/IsBsonType.php (hunk 1 above)
Comment (jmikola) [reply tests/UnifiedSpecTests/Constraint/IsBsonType.php]: This should now be fixed. If my new approach looks good, free to backport to v1.6. IIRC, that's when DocumentMatchConstraint got `$$type` support.

File: composer.json [hunk 5]
```diff
@@ -26,7 +26,10 @@
         "files": [ "src/functions.php" ]
     },
     "autoload-dev": {
-        "psr-4": { "MongoDB\\Tests\\": "tests/" }
+        "psr-4": { "MongoDB\\Tests\\": "tests/" },
+        "// Manually include assertion functions for PHPUnit 8.x and earlier ":"",
+        "// See: https://github.com/sebastianbergmann/phpunit/issues/3746 ":"",
```
Comment (jmikola) [inline composer.json:31]: Picked up this suggestion from https://gist.github.com/MoOx/5271067

File: tests/UnifiedSpecTests/Context.php [hunk 6]
```diff
@@ -0,0 +1,395 @@
...
+    private static function assertHasOnlyKeys($arrayOrObject, array $keys)
+    {
+        assertThat($arrayOrObject, logicalOr(isType('array'), isType('object')));
+        $diff = array_diff_key((array) $arrayOrObject, array_fill_keys($keys, 1));
+        assertEmpty($diff, 'Unsupported keys: ' . implode(',', array_keys($diff)));
+    }
+
+    private function createClient(stdClass $o) : Client
+    {
+        self::assertHasOnlyKeys($o, ['id', 'uriOptions', 'useMultipleMongoses', 'observeEvents', 'ignoreCommandMonitoringEvents']);
```
Comment (alcaeus) [inline tests/UnifiedSpecTests/Context.php:168]: Nice way to be defensive about what keys we support. Would it make sense to expand the error message a bit to give pointers about where the unsupported keys occur? So maybe "Unsupported keys in client entity: ..." as opposed to the current "Unsupported keys: ...".

File: tests/UnifiedSpecTests/Context.php [hunk 7]
```diff
@@ -0,0 +1,395 @@
...
+            } else {
+                $uri = self::removeMultipleMongoses($uri);
+            }
+        }
+
+        $uriOptions = [];
+
+        if (isset($o->uriOptions)) {
+            assertInternalType('object', $o->uriOptions);
+            /* TODO: If readPreferenceTags is set, assert it is an array of
```
Comment (alcaeus) [inline tests/UnifiedSpecTests/Context.php:190]: Leaving a note for future me to check back.

File: tests/UnifiedSpecTests/Context.php [hunk 8]
```diff
@@ -0,0 +1,395 @@
...
+    private function createCollection(stdClass $o) : Collection
+    {
+        self::assertHasOnlyKeys($o, ['id', 'database', 'collectionName', 'collectionOptions']);
+
+        $collectionName = $o->collectionName ?? null;
+        $database = $o->database ?? null;
+
+        assertInternalType('string', $collectionName);
+        assertInternalType('string', $database);
+        assertArrayHasKey($database, $this->entityMap);
```
Comment (alcaeus) [inline tests/UnifiedSpecTests/Context.php:215]: Just wanted to point out that the entity map does this check already, but also provides a more helpful error message ("No entity is defined for ...") as opposed to a simple ("Failed asserting that an array has the key ...").

I think we can remove the check on the entity map here (and in all occurrences that follow) and defer to the map itself to perform the right checks.

File: tests/UnifiedSpecTests/UnifiedSpecTest.php [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+     * @param string $schemaVersion
+     * @return boolean
+     */
+    private function isSchemaVersionSupported($schemaVersion)
+    {
+        if (version_compare($schemaVersion, '1.0', '>=') && version_compare($schemaVersion, '1.1', '<')) {
+            return true;
+        }
+
+        return false;
```
Comment (alcaeus) [inline tests/UnifiedSpecTests/UnifiedSpecTest.php:232]: Suggesting a shorter alternative similar to `isShardedCluster`:

```suggestion
        return version_compare($schemaVersion, '1.0', '>=') && version_compare($schemaVersion, '1.1', '<');
```

**Review by alcaeus — COMMENTED** [review]
Initial review. As per request, I've skipped reviewing the `ExpectedError`, `ExpectedResult`, and `Operation` classes.

File: tests/UnifiedSpecTests/CollectionData.php [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+     * @param Client $client
+     */
+    public function prepare(Client $client)
+    {
+        $database = $client->selectDatabase(
+            $this->databaseName,
+            ['writeConcern' => new WriteConcern(WriteConcern::MAJORITY)]
+        );
+
+        $database->dropCollection($this->collectionName);
```
[the remark this answers is not in the document]
Comment (jmikola) [reply tests/UnifiedSpecTests/CollectionData.php:56]: That commit isn't associated with any branch. Do you have more context for that?

Collections are only dropped if they appear in `initialData`, so this seems like something I may want to specify in the spec itself (e.g. "SHOULD NOT use admin collections in `initialData`), since nothing is currently stopping a test from doing so.

[the remark this answers is not in the document]
Comment (alcaeus) [reply tests/UnifiedSpecTests/CollectionData.php:56]: GitHub seems to mess up the link to the commit in the PR. The commit was part of #780, but viewing the commit standalone shows it as not being part of a branch due to squashed commits. Not sure I like that part about commit squashing...

[the remark this answers is not in the document]
Comment (alcaeus) [reply tests/UnifiedSpecTests/CollectionData.php:56]: For more context, the problem was crud/db-aggregate.json, which uses the admin database to run an aggregate with the `$listLocalSessions` pipeline stage. I'm not aware of a limitation that would require running this pipeline stage against the admin database, so adding a limitation in the spec should also suffice.

[the remark this answers is not in the document]
Comment (jmikola) [reply tests/UnifiedSpecTests/CollectionData.php:56]: Ok, that shouldn't be a problem then. Merely creating an enttiy for the admin database won't have the unified test runner drop it, and a test for `$listLocalSessions` wouldn't need to use `initialData` on an admin collection (or any collection for that matter).

What you're referring to was likely the same thing I ran into while writing tests for Atlas Data Lake, as the test runner attempted to drop the collection under test by default (which unmapped the data in `mongohoused`).

[the remark this answers is not in the document]
Comment (alcaeus) [reply tests/UnifiedSpecTests/CollectionData.php:56]: Ah, yes, that's the problem then. In that case, nothing to do here 👍

File: tests/UnifiedSpecTests/CollectionData.php [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+        $mi = new MultipleIterator(MultipleIterator::MIT_NEED_ANY);
+        $mi->attachIterator(new ArrayIterator($this->documents));
+        $mi->attachIterator(new IteratorIterator($cursor));
+
+        foreach ($mi as $i => $documents) {
+            list($expectedDocument, $actualDocument) = $documents;
+            assertNotNull($expectedDocument);
+            assertNotNull($actualDocument);
+
+            /* Disable extra root keys and operators when matching, which is
```
[the remark this answers is not in the document]
Comment (jmikola) [reply tests/UnifiedSpecTests/CollectionData.php:95]: Rephrased as:

> Prohibit extra root keys and disable operators to enforce exact matching of documents. Key order variation is still allowed.

File: tests/UnifiedSpecTests/CollectionData.php [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+
+        assertInternalType('array', $o->documents);
+        assertContainsOnly('object', $o->documents);
+        $this->documents = $o->documents;
+    }
+
+    /**
+     * Prepare collection state for "initialData".
+     *
+     * @param Client $client
```
[the remark this answers is not in the document]
Comment (jmikola) [reply tests/UnifiedSpecTests/CollectionData.php:47]: Removed a number of doc blocks but kept some where the comment wasn't entirely obvious. In this case, I renamed the method to refer to "initialData".

File: tests/UnifiedSpecTests/Constraint/Matches.php [hunk 13]
```diff
@@ -0,0 +1,442 @@
...
+        $name = self::getOperatorName($operator);
+
+        if ($name === '$$type') {
+            assertThat(
+                $operator['$$type'],
+                logicalOr(isType('string'), logicalAnd(isInstanceOf(BSONArray::class), containsOnly('string'))),
+                '$$type requires string or string[]'
+            );
+
+            $types = is_string($operator['$$type']) ? [$operator['$$type']] : $operator['$$type'];
```
[the remark this answers is not in the document]
Comment (jmikola) [reply tests/UnifiedSpecTests/Constraint/Matches.php:245]: Good suggestion. This was leftover from before the assertions were added.

File: tests/UnifiedSpecTests/Constraint/Matches.php [hunk 14]
```diff
@@ -0,0 +1,451 @@
...
+                self::failAt(sprintf('%s is not an expected BSON type: %s', $this->exporter()->shortenedExport($actual), implode(', ', $types)), $keyPath);
+            }
+
+            return;
+        }
+
+        if ($name === '$$matchesEntity') {
+            assertInternalType('string', $operator['$$matchesEntity'], '$$matchesEntity requires string');
+
+            $this->assertMatches(
```
[the remark this answers is not in the document]
Comment (jmikola) [reply tests/UnifiedSpecTests/Constraint/Matches.php:264]: Will give this some more thought. I think it'd require catching and re-throwing the exception to add an additional prefix/suffix.

Your question made me realize that the entity value itself might be a document that contains more operators, which is an edge case but something worth clarifying in the spec. I can't think of a reason to support additional operators when recursing here.

File: tests/UnifiedSpecTests/Context.php (hunk 6 above)
Comment (jmikola) [reply tests/UnifiedSpecTests/Context.php:168]: The stack trace of the assertion failure should tell us the exact context, so I don't think it's worth the trouble. In the case of `prepareCollectionOrDatabaseOptions` we wouldn't know the exact database or collection entity anyway.

Generally we should never hit these errors unless we're executing test files that don't comply with the JSON schema, since I used `additionalProperties: false` throughout.

File: tests/UnifiedSpecTests/Context.php (hunk 8 above)
Comment (jmikola) [reply tests/UnifiedSpecTests/Context.php:215]: Also removing the `assertArrayNotHasKey` in `createEntities`, since EntityMap enforces that we cannot overwrite entities.

File: tests/UnifiedSpecTests/UnifiedSpecTest.php (hunk 9 above)
Comment (jmikola) [reply tests/UnifiedSpecTests/UnifiedSpecTest.php:232]: SGTM. This was written with the expectation that we might one day support multiple ranges (e.g. 1.0-1.5, 2.0-2.3), but that's a long way off.

File: tests/UnifiedSpecTests/Context.php (hunk 6 above)
Comment (alcaeus) [reply tests/UnifiedSpecTests/Context.php:164]: That makes sense. In that case, I'm happy to leave this as is.

File: tests/UnifiedSpecTests/Constraint/Matches.php (hunk 14 above)
[the remark this answers is not in the document]
Comment (jmikola) [reply tests/UnifiedSpecTests/Constraint/Matches.php:259]: Left a TODO item in the code so we can resolve this.

**Comment by jmikola** [comment]
Includes patch from https://github.com/mongodb/mongo-php-library/pull/790 for CreateIndexes.

Also syncs spec tests with https://github.com/mongodb/specifications/pull/846, all of which are currently passing locally on replica set and sharded clusters. We'll see what Travis and Evergreen say.

File: tests/UnifiedSpecTests/UnifiedSpecTest.php [hunk 15]
```diff
@@ -0,0 +1,346 @@
...
+        $topology = $this->getCachedTopology();
+
+        foreach ($runOnRequirements as $o) {
+            $runOnRequirement = new RunOnRequirement($o);
+            if ($runOnRequirement->isSatisfied($serverVersion, $topology)) {
+                return;
+            }
+        }
+
+        $this->markTestSkipped(sprintf('Server version "%s" and topology "%s" do not meet test requirements', $serverVersion, $topology));
```
Comment (alcaeus) [inline tests/UnifiedSpecTests/UnifiedSpecTest.php:226]: I'll see if I can print the requirements in the message as well. I think it would be helpful, WDYT?

File: tests/UnifiedSpecTests/UnifiedSpecTest.php [hunk 16]
```diff
@@ -0,0 +1,346 @@
...
+
+    /**
+     * Checks is a test format schema version is supported.
+     *
+     * @param string $schemaVersion
+     * @return boolean
+     */
+    private function isSchemaVersionSupported($schemaVersion)
+    {
+        return version_compare($schemaVersion, '1.0', '>=') && version_compare($schemaVersion, '1.1', '<');
```
Comment (alcaeus) [inline tests/UnifiedSpecTests/UnifiedSpecTest.php:291]: We could extract these to constants on the class itself to have the supported version declared at the top of the class.

File: tests/UnifiedSpecTests/Constraint/IsBsonType.php [hunk 17]
```diff
@@ -0,0 +1,210 @@
...
+
+    public static function any() : LogicalOr
+    {
+        return self::anyOf(...self::$types);
+    }
+
+    public static function anyOf(string ...$types) : Constraint
+    {
+        if (count($types) === 1) {
+            return new self(...$types);
```
[the remark this answers is not in the document]
Comment (jmikola) [reply tests/UnifiedSpecTests/Constraint/IsBsonType.php:90]: This PR was my first time getting to use the splat operator and I went all in.

File: tests/UnifiedSpecTests/Constraint/Matches.php [hunk 18]
```diff
@@ -0,0 +1,453 @@
...
+            return $bson;
+        }
+
+        /* Convert Int64 objects to integers on 64-bit platforms for
+         * compatibility reasons. */
+        if ($bson instanceof Int64 && PHP_INT_SIZE != 4) {
+            return (int) ((string) $bson);
+        }
+
+        /* TODO: Convert Int64 objects to integers on 32-bit platforms if they
```
[the remark this answers is not in the document]
Comment (jmikola) [reply tests/UnifiedSpecTests/Constraint/Matches.php:410]: It may actually never come up unless a test actually asserts that a cursor ID is zero and a 32-bit PHP platform tries to compare `0` with `Int64(0)`.

File: tests/UnifiedSpecTests/EventObserver.php [hunk 19]
```diff
@@ -0,0 +1,279 @@
...
+        $this->clientId = $clientId;
+        $this->context = $context;
+    }
+
+    /**
+     * @see https://www.php.net/manual/en/mongodb-driver-monitoring-commandsubscriber.commandfailed.php
+     */
+    public function commandFailed(CommandFailedEvent $event)
+    {
+        if (! $this->context->isActiveClient($this->clientId)) {
```
[the remark this answers is not in the document]
Comment (jmikola) [reply tests/UnifiedSpecTests/EventObserver.php:100]: Ah, I see what you mean. These can definitely share a single private method that just checks the class name.

File: tests/UnifiedSpecTests/ExpectedError.php [hunk 20]
```diff
@@ -0,0 +1,191 @@
...
+        'MaxTimeMSExpired' => 50,
+        'NoSuchTransaction' => 251,
+        'OperationNotSupportedInTransaction' => 263,
+        'WriteConflict' => 112,
+    ];
+
+    /** @var bool */
+    private $isError = false;
+
+    /** @var bool */
```
[the remark this answers is not in the document]
Comment (jmikola) [reply tests/UnifiedSpecTests/ExpectedError.php:44]: Ah, `?bool`, etc. Indeed, this was me doing the bare minimum to get `phpcs` to shut up.

On that note, is it worth requiring `@var` annotations in our tests? If these help your IDE I don't object.

File: tests/UnifiedSpecTests/Util.php [hunk 21]
```diff
@@ -0,0 +1,112 @@
...
+use function assertThat;
+use function implode;
+use function isType;
+use function logicalOr;
+
+final class Util
+{
+    public static function assertHasOnlyKeys($arrayOrObject, array $keys)
+    {
+        // TODO: replace isType('object') with instanceOf(stdClass::class)
```
[the remark this answers is not in the document]
Comment (jmikola) [reply tests/UnifiedSpecTests/Util.php:25]: My initial reply was going to be: I think `stdClass` is warranted for two reasons:

* In practice, this is only used to validate options after parsing test files.
* `stdClass` is the only object that we can trust to reliably cast as an array for the subsequent `array_diff_key`. Other objects might have private/protected properties.

Then, for I reminded myself that non-public properties still surface in an array cast:

```
>>> (array) new class { private $a; protected $b; public $c; function __construct() { $a=1;$b=2;$c=3; } }
=> [
     "\0class@anonymous\0/home/jmikola/.config/composer/vendor/psy/psysh/src/ExecutionLoopClosure.php(50) : eval()'d code:1$2a4\0a" => null,
     "\0*\0b" => null,
     "c" => null,
   ]
```

In that case, such objects would likely fail if they had those properties, since I don't expect we'd include those keys in the argument. No worries if you'd like to keep this as-is and just delete the TODO.

File: tests/UnifiedSpecTests/UnifiedSpecTest.php [hunk 22]
```diff
@@ -0,0 +1,346 @@
...
+        if ($this->hasFailed()) {
+            self::killAllSessions();
+        }
+
+        $this->failPointObserver->stop();
+        $this->failPointObserver->disableFailPoints();
+
+        /* TODO: Consider manually invoking gc_collect_cycles since each test is
+         * prone to create cycles (perhaps due to EntityMap), which can leak and
+         * prevent sessions from being released back into the pool. */
```
[the remark this answers is not in the document]
Comment (jmikola) [reply tests/UnifiedSpecTests/UnifiedSpecTest.php:80]: I only bumped into this while investigating the blocking transaction test issue, and that ultimately required `killAllSessions` as a workaround. I don't actually think this will have any effect on test execution, although there's some peace of mind in being able to trust that the EntityMap is totally dismantled. I imagine this will be more relevant when/if we start using Clients with persistence disabled, since we'd probably want to reap those after each test.

File: tests/UnifiedSpecTests/UnifiedSpecTest.php [hunk 23]
```diff
@@ -0,0 +1,346 @@
...
+        }
+
+        // Give Context unmodified URI so it can enforce useMultipleMongoses
+        $context = new Context(self::$internalClient, static::getUri(true));
+
+        if (isset($createEntities)) {
+            $context->createEntities($createEntities);
+        }
+
+        // TODO handle distinct commands in sharded transactions
```
[the remark this answers is not in the document]
Comment (jmikola) [reply tests/UnifiedSpecTests/UnifiedSpecTest.php:119]: Basically copy in https://github.com/mongodb/mongo-php-library/blob/1.7.1/tests/SpecTests/TransactionsSpecTest.php#L318. Rather than run that for _each_ test, I think we can be a bit smarter and have the function quickly check if a sharded environment is being used _and_ scan the `operations` array for both `startTransaction` and `distinct` (order doesn't really matter, since it's just making an educated guess).

File: tests/UnifiedSpecTests/Constraint/Matches.php (hunk 18 above)
[the remark this answers is not in the document]
Comment (alcaeus) [reply tests/UnifiedSpecTests/Constraint/Matches.php:410]: Sounds good 👍

File: tests/UnifiedSpecTests/EventObserver.php (hunk 19 above)
[the remark this answers is not in the document]
Comment (alcaeus) [reply tests/UnifiedSpecTests/EventObserver.php:100]: Hehe, I know the feeling. FYI, we'll have to use `get_class($event)` to check if the event is observed until PHP 8. From then on, we can use `$event::class` :)

File: tests/UnifiedSpecTests/UnifiedSpecTest.php (hunk 15 above)
Comment (jmikola) [reply tests/UnifiedSpecTests/UnifiedSpecTest.php:226]: I'm used to opening tests based on the name, so up to you. I imagine you'd either want to spread it across multiple lines or `json_encode()` the entire thing for single-line output.

File: tests/UnifiedSpecTests/ExpectedError.php (hunk 20 above)
[the remark this answers is not in the document]
Comment (alcaeus) [reply tests/UnifiedSpecTests/ExpectedError.php:44]: They do indeed help the IDE, and they will also help with phpstan or psalm if I ever get around to introducing it.

File: tests/UnifiedSpecTests/Util.php (hunk 21 above)
[the remark this answers is not in the document]
Comment (alcaeus) [reply tests/UnifiedSpecTests/Util.php:25]: Ah, in that case, let me completely reverse my opinion and suggest we change this to `instanceOf(stdClass::class)` as soon as it's safe to do so

File: tests/UnifiedSpecTests/UnifiedSpecTest.php (hunk 22 above)
[the remark this answers is not in the document]
Comment (alcaeus) [reply tests/UnifiedSpecTests/UnifiedSpecTest.php:80]: Yep. Considering there already was some funkiness going on and with the segmentation fault appearing every once in a while due to the clusterTime being double-freed, I'm all good for using `gc_collect_cycles` as a precaution

File: tests/UnifiedSpecTests/UnifiedSpecTest.php (hunk 15 above)
Comment (alcaeus) [reply tests/UnifiedSpecTests/UnifiedSpecTest.php:226]: In that case, let's keep it as-is for the time being. 👍

**Review by alcaeus — APPROVED** [review]
